### PR TITLE
DQN

### PR DESCRIPTION
**Pull request by vkhizanov**

### CODE REVIEW

File: relaax/algorithms/dqn/agent.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+class Agent(object):
+    def __init__(self, parameter_server, metrics):
+        self.ps = parameter_server
+        self.metrics = metrics
+
+    # environment is ready and
+    # waiting for agent to initialize
+    def init(self, exploit=False, hogwild_update=False):
+        self.episode = dqn_episode.DQNEpisode(self.ps, self.metrics, exploit, hogwild_update)
```
Comment (4SkyNet) [inline relaax/algorithms/dqn/agent.py]: better to use like a: 
```python
dqn_episode.Trainer(self.ps, self.metrics, exploit, hogwild_update)
```
wrt to new conception of `Trainers`

File: relaax/algorithms/dqn/dqn_model.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        hidden = layer.GenericLayers(layer.Flatten(input),
+                                     [dict(type=layer.Dense, size=size, activation=layer.Activation.Tanh) for size in cfg.config.hidden_sizes])
+
+        output = layer.Dense(hidden, cfg.config.output.action_size)
+
+        layers = [input, hidden, output]
+
+        self.ph_state = input.ph_state
+        self.output = output
+        self.weights = layer.Weights(*layers)
```
Comment (4SkyNet) [inline relaax/algorithms/dqn/dqn_model.py]: or immediate like that:
```python
self.weights = layer.Weights(input, hidden, output)
```

File: relaax/algorithms/dqn/dqn_model.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        sg_target_network = Network()
+
+        sg_get_action = Action()
+
+        if cfg.config.optimizer == 'Adam':
+            sg_optimizer = graph.AdamOptimizer(cfg.config.initial_learning_rate)
+        else:
+            raise NotImplementedError
+
+        sg_loss = loss.DQNLoss(sg_network.output, cfg.config.output, cfg.config.double_dqn, cfg.config.rewards_gamma)
```
Comment (4SkyNet) [inline relaax/algorithms/dqn/dqn_model.py]: I think it's better to pass only `cfg.config`

File: relaax/algorithms/dqn/dqn_model.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        sg_gradients_calc = layer.Gradients(sg_network.weights, loss=sg_loss)
+        sg_gradients_apply = layer.Gradients(sg_network.weights, optimizer=sg_optimizer)
+
+        sg_update_target_weights = graph.TfNode([tf.assign(variable, value) for variable, value in utils.Utils.izip(sg_target_network.weights.node, sg_network.weights.node)])
```
Comment (4SkyNet) [inline relaax/algorithms/dqn/dqn_model.py]: I wrote `graph.AssignWeights(acceptor, donor)` in new `graph` file at `pixie` branch

File: relaax/algorithms/dqn/dqn_model.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+import tensorflow as tf
+
+from relaax.common.algorithms import subgraph
+from relaax.common.algorithms.lib import graph
+from relaax.common.algorithms.lib import layer
+from relaax.common.algorithms.lib import loss
+from relaax.common.algorithms.lib import utils
+
+from . import dqn_config as cfg
+from .lib.dqn_utils import Action
```
Comment (4SkyNet) [inline relaax/algorithms/dqn/dqn_model.py]: I guess it should be `actor` -> cuz we mostly use this definition

File: relaax/algorithms/dqn/dqn_model.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+        self.op_initialize = self.Op(sg_initialize)
+
+
+class AgentModel(subgraph.Subgraph):
+    def build_graph(self):
+        sg_network = Network()
+        sg_target_network = Network()
+
+        sg_get_action = Action()
```
Comment (4SkyNet) [inline relaax/algorithms/dqn/dqn_model.py]: or:
```python
sg_get_action = sg_network.actor
```

File: relaax/algorithms/dqn/lib/dqn_episode.py [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+from . import dqn_utils
+
+import random
+
+
+logger = logging.getLogger(__name__)
+profiler = profiling.get_profiler(__name__)
+
+
+class DQNEpisode(object):
```
Comment (4SkyNet) [inline relaax/algorithms/dqn/lib/dqn_episode.py]: `Trainer` =)

File: relaax/algorithms/dqn/lib/dqn_episode.py [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+    def __init__(self, parameter_server, metrics, exploit, hogwild_update):
+        self.ps = parameter_server
+        self.metrics = metrics
+
+        self.session = session.Session(dqn_model.AgentModel())
+        self.session.op_initialize()
+
+        self.replay_buffer = dqn_utils.ReplayBuffer(dqn_config.config.replay_buffer_size)
+        self.observation = dqn_utils.DQNObservation()
```
Comment (4SkyNet) [inline relaax/algorithms/dqn/lib/dqn_episode.py]: I moved the `observation` class to `common` cuz it does the same for all algorithms -> just stack consecutive states.
```python
from relaax.common.algorithms.lib import observation
...
self.observation = observation.Observation()
```

File: relaax/algorithms/dqn/lib/dqn_episode.py [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+        self.queue = None
+
+    @profiler.wrap
+    def begin(self):
+        self.get_action()
+
+    @profiler.wrap
+    def step(self, reward, state, terminal):
+        self.local_step += 1
+        self.last_target_weights_update += 1
```
Comment (4SkyNet) [inline relaax/algorithms/dqn/lib/dqn_episode.py]: good

File: relaax/algorithms/dqn/lib/dqn_episode.py [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+        q_next = self.session.op_get_q_value(state=batch["next_state"])
+
+        feeds = dict(state=batch["state"],
+                     reward=batch["reward"],
+                     action=batch["action"],
+                     terminal=batch["terminal"],
+                     q_next_target=q_next_target,
+                     q_next=q_next)
+
+        gradients = self.session.op_compute_gradients(**feeds)
```
Comment (4SkyNet) [inline relaax/algorithms/dqn/lib/dqn_episode.py]: nice

File: relaax/common/algorithms/lib/layer.py [hunk 11]
```diff
@@ -151,7 +151,7 @@ def build_graph(self, input, descs=None):
 
         if np.prod(input_shape) == 0:
             input_shape = [1]
-        shape = [None] + input_shape + [input.history]
+        shape = [None] + input_shape + ([input.history] if hasattr(input, 'history') else [])
```
Comment (4SkyNet) [inline relaax/common/algorithms/lib/layer.py]: Good

File: relaax/templates/configs/dqn.yaml [hunk 12]
```diff
@@ -0,0 +1,41 @@
+---
+# generic DA3C configuration
+
+algorithm:
+  name: dqn                       # short name for algorithm to load
+
+  input:
+    shape: [4]                    # shape of input state
+    history: 1                    # number of consecutive states to stack
```
Comment (4SkyNet) [inline relaax/templates/configs/dqn.yaml]: I hate `history` cuz it isn't so straightforward -> @StanislavVolodarskiy can we rename it in some more appropriate way for user experience? ...I think we should to `rewrite` only `input` layer to support this renaming

File: relaax/templates/configs/dqn.yaml [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+    continuous: false             # set to true to switch to continuous action space
+    action_size: 2                # action size for the given environment
+
+  double_dqn: true
+
+  hidden_sizes: [64]              # list of dense layers sizes, for ex. [128, 64]
+  batch_size: 32                  # maximum batch size, which need to accumulate for one update
+
+  max_global_step: 150000         # amount of maximum global steps to pass through the training
+  start_sample_step: 2000
```
Comment (4SkyNet) [inline relaax/templates/configs/dqn.yaml]: It's good to add some description to these `params` and how we can tune them if it possible to handle the process of learning which could be a `+` for user-exp

File: relaax/algorithms/dqn/dqn_model.py [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+    def build_graph(self):
+        sg_network = Network()
+        sg_target_network = Network()
+
+        sg_get_action = Actor()
+
+        if cfg.config.optimizer == 'Adam':
+            sg_optimizer = graph.AdamOptimizer(cfg.config.initial_learning_rate)
+        else:
+            raise NotImplementedError
```
Comment (StanislavVolodarskiy) [inline relaax/algorithms/dqn/dqn_model.py]: Are you going to add other optimizers later?

File: relaax/algorithms/dqn/lib/dqn_trainer.py [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+
+import random
+
+
+logger = logging.getLogger(__name__)
+profiler = profiling.get_profiler(__name__)
+
+
+class Trainer(object):
+    def __init__(self, parameter_server, metrics, exploit, hogwild_update):
```
Comment (StanislavVolodarskiy) [inline relaax/algorithms/dqn/lib/dqn_trainer.py]: hogwild_update is not used. And self.queue and self.do_task. Please remove them if you're not going to support hogwild update in nearest future.

File: relaax/common/algorithms/lib/layer.py (hunk 11 above)
Comment (StanislavVolodarskiy) [reply relaax/common/algorithms/lib/layer.py]: I'm in doubt about this change. Lets test it with history absent in app.yaml.

File: relaax/algorithms/dqn/dqn_model.py [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+
+import tensorflow as tf
+
+
+class Network(subgraph.Subgraph):
+    def build_graph(self):
+        input = layer.Input(cfg.config.input)
+
+        hidden = layer.GenericLayers(layer.Flatten(input),
+                                     [dict(type=layer.Dense, size=size, activation=layer.Activation.Tanh) for size in cfg.config.hidden_sizes])
```
Comment (StanislavVolodarskiy) [inline relaax/algorithms/dqn/dqn_model.py]: Line looks too long. Please install flake8 (pip install flake8) and run it on modified/new files.

File: relaax/algorithms/dqn/dqn_model.py [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+
+        weights = [input, hidden]
+
+        if cfg.config.dueling_dqn:
+            v_input, a_input = tf.split(hidden.node, [cfg.config.hidden_sizes[-1] // 2, cfg.config.hidden_sizes[-1] // 2], axis=1)
```
Comment (StanislavVolodarskiy) [inline relaax/algorithms/dqn/dqn_model.py]: Do you assume last hidden size is even?
Please consider situation when hidden_size list is empty. This could be case for extremely simple environments.

File: relaax/algorithms/dqn/dqn_model.py [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+
+            v_input = graph.TfNode(v_input)
+            a_input = graph.TfNode(a_input)
+
+            v_output = layer.Dense(v_input, 1)
+            a_output = layer.Dense(a_input, cfg.config.output.action_size)
+
+            output = graph.TfNode(tf.add(v_output.node, tf.subtract(a_output.node, tf.reduce_mean(a_output.node, axis=1, keep_dims=True))))
```
Comment (StanislavVolodarskiy) [inline relaax/algorithms/dqn/dqn_model.py]: This is matter of preference. I'd change tf.add and tf.subtract to + and - respectively.

File: relaax/algorithms/dqn/dqn_model.py [hunk 19]
```diff
@@ -0,0 +1,120 @@
...
+
+            weights.extend([v_output, a_output])
+        else:
+            output = layer.Dense(hidden, cfg.config.output.action_size)
```
Comment (StanislavVolodarskiy) [inline relaax/algorithms/dqn/dqn_model.py]: It seems like cfg is always followed by config. Lets import config directly.

File: relaax/algorithms/dqn/lib/dqn_trainer.py [hunk 20]
```diff
@@ -0,0 +1,131 @@
...
+from relaax.common import profiling
+from relaax.server.common import session
+from relaax.common.algorithms.lib import utils
+from relaax.common.algorithms.lib import observation
+
+from .. import dqn_config
+from .. import dqn_model
+from . import dqn_utils
+
+import random
```
Comment (StanislavVolodarskiy) [inline relaax/algorithms/dqn/lib/dqn_trainer.py]: Lets import standart modules before thrid-party ones and third-party ones before relaax ones.

File: relaax/algorithms/dqn/lib/dqn_trainer.py [hunk 21]
```diff
@@ -0,0 +1,131 @@
...
+        if state is not None:
+            self.metrics.histogram('state', state)
+
+        if reward is not None:
+            self.push_experience(reward, state, terminal)
+        else:
+            self.observation.add_state(state)
+
+        if terminal:
+            self.observation.add_state(None)
```
Comment (StanislavVolodarskiy) [inline relaax/algorithms/dqn/lib/dqn_trainer.py]: Two self.observation.add_state in row looks suspicious. Are you going to add_state twice in case of terminal?

File: relaax/algorithms/dqn/lib/dqn_utils.py [hunk 22]
```diff
@@ -0,0 +1,48 @@
...
+        else:
+            weights = map(lambda x: math.pow(x, self._alpha), range(1, len(self._replay_memory) + 1)) if self._weights is None else self._weights
+            return random.choices(self._replay_memory,
+                                  weights=weights,
+                                  k=size)
+
+    def append(self, value):
+        self._replay_memory.append(value)
+        if len(self._replay_memory) == self._replay_memory.maxlen:
+            self._weights = list(map(lambda x: math.pow(x, self._alpha), range(1, self._replay_memory.maxlen + 1)))
```
Comment (StanislavVolodarskiy) [inline relaax/algorithms/dqn/lib/dqn_utils.py]: List comprehension looks better in this case.

File: relaax/algorithms/dqn/agent.py [hunk 23]
```diff
@@ -0,0 +1,29 @@
...
+
+
+class Agent(object):
+    def __init__(self, parameter_server, metrics):
+        self.ps = parameter_server
+        self.metrics = metrics
+
+    # environment is ready and
+    # waiting for agent to initialize
+    def init(self, exploit=False, hogwild_update=False):
```
Comment (StanislavVolodarskiy) [inline relaax/algorithms/dqn/agent.py]: hogwild_update is not used. And exploit.
About exploit: Are you going to support it?

File: relaax/algorithms/dqn/dqn_model.py [hunk 24]
```diff
@@ -0,0 +1,133 @@
...
+
+        hidden = layer.GenericLayers(layer.Flatten(input),
+                                     [dict(type=layer.Dense, size=size, activation=layer.Activation.Tanh)
+                                      for size in config.hidden_sizes])
+
+        weights = [input, hidden]
+
+        if config.dueling_dqn:
+            if config.hidden_sizes:
+                assert config.hidden_sizes[-1] % 2 == 0
```
Comment (StanislavVolodarskiy) [inline relaax/algorithms/dqn/dqn_model.py]: hidden_sizes is kind of user input. Lets check it in dqn_config.py showing user friendly error message if needed.

File: relaax/algorithms/dqn/lib/dqn_utils.py [hunk 25]
```diff
@@ -0,0 +1,53 @@
...
+            weights = map(lambda x: math.pow(x, self._alpha),
+                          range(1, len(self._replay_memory) + 1)) if self._weights is None else self._weights
+            return random.choices(self._replay_memory,
+                                  weights=weights,
+                                  k=size)
+
+    def append(self, value):
+        self._replay_memory.append(value)
+        if len(self._replay_memory) == self._replay_memory.maxlen:
+            self._weights = [math.pow(x, self._alpha) for x in range(1, self._replay_memory.maxlen + 1)]
```
Comment (StanislavVolodarskiy) [inline relaax/algorithms/dqn/lib/dqn_utils.py]: Lets create self._weigths once. One use it slicing in case of incomplete queue.

File: relaax/common/algorithms/lib/loss.py [hunk 26]
```diff
@@ -38,6 +38,27 @@ def build_graph(self, actor, critic, cfg):
         return self.policy_loss + self.value_loss
 
 
+class DQNLoss(subgraph.Subgraph):
+    def build_graph(self, q_network, cfg):
+        self.ph_reward = tf.placeholder(tf.float32, [None])
```
Comment (StanislavVolodarskiy) [inline relaax/common/algorithms/lib/loss.py]: Our policy is not passing configuration as parameter. Import configuration module here and use it directly.

File: relaax/common/algorithms/lib/loss.py [hunk 27]
```diff
@@ -38,6 +38,27 @@ def build_graph(self, actor, critic, cfg):
...
+
+        action_one_hot = tf.one_hot(self.ph_action, cfg.output.action_size)
+        q_action = tf.reduce_sum(tf.multiply(q_network.node, action_one_hot), axis=1)
+
+        if cfg.double_dqn:
+            q_max = tf.reduce_sum(tf.multiply(self.ph_q_next_target, tf.one_hot(tf.argmax(self.ph_q_next, axis=1), cfg.output.action_size)), axis=1)
+        else:
+            q_max = tf.reduce_max(self.ph_q_next_target, axis=1)
+
+        y = tf.add(self.ph_reward, tf.multiply(tf.cast(tf.subtract(1, self.ph_terminal), tf.float32), tf.scalar_mul(cfg.rewards_gamma, q_max)))
```
Comment (StanislavVolodarskiy) [inline relaax/common/algorithms/lib/loss.py]: Lets change tf.add, tf.multiply to +, * if do not have any objections.
